### PR TITLE
refactor(data): make input modality first-class and Data-owned (refs #253)

### LINKED ISSUE #253: Generalize Data layer beyond image/file-centric assumptions (TaskFamily follow-up)
## Context

Follow-up to the **TaskFamily refactor** (collapses the ~18 scattered `task_kind` branches into per-family strategy objects). That refactor introduces `TaskFamily.load_labels` as the per-family label-loading hook, but deliberately leaves the surrounding `Data` layer untouched to stay in scope. (Note: there is no `load_inputs` hook yet — see Deferred / future work below.)

This issue tracks generalizing the `Data` layer itself.

## Problem

The `Data` module hard-assumes **image/file** inputs:

- **Sample discovery** is filesystem/image-folder based (ImageFolder-style layouts, file globbing).
- **`sample_ids`** are effectively filenames/stems.
- **`InputSpec` / `input_metadata`** carries image-shaped fields (dims).
- **`preprocessing`** is resize/crop-shaped (`Weights.transforms()`, CenterCrop).

These assumptions block non-image task families surfaced during the TaskFamily design pressure-test:

- **seq2seq** (translation/summarization): inputs are token sequences (variable length + attention masks), not image files; labels are reference sequences; "preprocessing" means tokenization, not resize.
- **segmentation / others** similarly strain the file/image model.

## Scope (this issue)

- Decouple sample discovery + `sample_ids` from filesystem/image layout.
- Make `InputSpec` / `input_metadata` task-family-general (not image-dim-only).
- Generalize or cleanly opt out of the image-shaped `preprocessing` concept per family.

## Out of scope

- The TaskFamily refactor itself (separate spec/issue).
- Adding any concrete new family (segmentation/seq2seq) — this is infra only.

## Notes

- `allows_preprocessing -> False` is the current clean opt-out for non-image families; this issue is about making the positive path general, not just the opt-out.
- Pre-1.0: backward-compat breaks are acceptable (schema/API).

## Deferred / future work (post-design)

- **`load_inputs` hook.** The input-side mirror of `load_labels`, for **torch-currency** non-image families (seq2seq, segmentation) whose source read cannot go through the dense-tensor seam (`adapt_loaded_inputs`). Not added here: with no such family yet, the signature is a guess. Its first consumer (seq2seq) also needs a multi-tensor forward + explainer that is not designed — see #101. Land `load_inputs` with that consumer. Today inputs are co-owned (`Data` reads bytes → dense tensor; the family reshapes via `adapt_loaded_inputs`); `load_labels` is family-owned end-to-end. `load_inputs` only earns its place when the read itself must be family-owned.
- **Non-torch backends are out of scope.** trees / scikit / TensorFlow break the `torch.Tensor` / `nn.Module` currency in `models/backend.py` + `transparency/`, not the `Data` layer. They belong to #209 (drop the `nn.Module` bridge) and #246 (tree backend / TreeSHAP).

This issue (input axis: "data isn't always an image") and #209 (model axis: "model isn't always a torch `nn.Module`") are independent — no single consumer needs both co-designed. Cross-ref: #209, #101, #246.

### PR DESCRIPTION
**Pull request by stanlrt**
## What

Safe core of #253: make input **modality** (image vs tabular) first-class and `Data`-owned, so metadata inference reads a recorded tag instead of re-globbing the filesystem, and the duplicated extension literals collapse to one home.

- `data/types.py` — new `InputModality` StrEnum (`image`/`tabular`) + `MODALITY_EXTENSIONS` canonical map (single source of truth for extensions).
- `data/data.py` — `_load_data` tags every successful return with its modality; `Data.input_modality` set in `__init__`; module constants now derive from the map.
- `data/metadata.py` — `infer_data_input_metadata` reads the recorded modality first; the source-path sniff is demoted to a fallback for callers with no recorded modality. Layout strings (`NCHW`/`(B,F)`) unchanged.
- `docs/contributor/modules/data.md` — documents the concept (contributor audience).

## Why

The modality was already determined inside `Data._load_data` but discarded; `metadata.py` then re-derived it by re-globbing the filesystem, duplicating the extension sets and emitting hardcoded layout strings. This removes the image-centric guessing in a layer that already knows the answer, and gives a future non-image family a clean seam.

## Behaviour

Internal refactor. Output is byte-identical for today's image/tabular inputs (the sniff fallback preserves direct/test callers). Not a breaking change.

## Out of scope (deferred, recorded in #253)

- `load_inputs` hook — needs a real torch-currency non-image family (seq2seq/segmentation); signature blocked on the multi-tensor forward + explainer (#101).
- Generalized preprocessing positive path; discovery/`sample_ids` registry.
- Non-torch backends (trees/scikit/TF) — the model axis, tracked in #209 / #246.

## Tests

- New `test_types.py`; modality + non-sniffable-source + fallback + demo-sample cases across `test_data_class.py` / `test_metadata.py`.
- `uv run pytest src/raitap/data/tests` → **160 passed**.
- `ruff check` + `ruff format` clean; `pyright` **0 errors**.

Refs #253 (safe core; remaining scope stays open).

## Checklist

- [x] **CI** — Required checks are green
- [x] **Breaking changes** — Not a breaking change (internal refactor, behaviour byte-identical for image/tabular); no `!` in the title.
- [x] **Contributor guide** — I’ve read [**Pull requests and commit messages**](https://caiivs.github.io/raitap/contributor/pull-requests.html) before requesting review.

## Optional

- [x] **Issue** _(optional)_ — Refs #253.
- [x] **Docs** _(optional)_ — `docs/contributor/modules/data.md` updated.
- [x] **Tests** _(optional)_ — New/updated tests cover the change (160 passed).

### CODE REVIEW

File: src/raitap/data/data.py [hunk 1]
```diff
@@ -57,7 +60,8 @@ def __init__(
         self.task_kind = task_kind
         self.tensor: torch.Tensor | DetectionInputs
         family = resolve_task_family(task_kind)
-        raw_tensor, self.sample_ids = self._load_data(
+        self.input_modality: InputModality | None
+        raw_tensor, self.sample_ids, self.input_modality = self._load_data(
             cfg,
```
[the remark this answers is not in the document]
Comment (stanlrt) [reply src/raitap/data/data.py:65]: Fixed in 39c8cea — narrowed to `self.input_modality: InputModality` (every `_load_data` success path returns a concrete member; errors raise).

File: src/raitap/data/data.py [hunk 2]
```diff
@@ -72,7 +76,7 @@ def _load_data(
         cfg: AppConfig,
         *,
         resolved_preprocessing: ResolvedPreprocessing | None = None,
-    ) -> tuple[torch.Tensor | DetectionInputs, list[str] | None]:
+    ) -> tuple[torch.Tensor | DetectionInputs, list[str] | None, InputModality]:
         """
         Load data from a specified source into a raw tensor.
```
[the remark this answers is not in the document]
Comment (stanlrt) [reply src/raitap/data/data.py:82]: Fixed in 39c8cea — `_load_data` `Returns:` now documents the `(tensor, sample_ids, input_modality)` tuple.
